core/capabilities/webapi/trigger.go

```diff
@@ -1,18 +1,268 @@
+type webapiTrigger struct {
+	allowedSenders map[string]bool
+	allowedTopics  map[string]bool
+	ch             chan<- capabilities.TriggerResponse
+	config         TriggerConfig
+	rateLimiter    *common.RateLimiter
+}
+
+type triggerConnectorHandler struct {
+	services.StateMachine
+
+	capabilities.CapabilityInfo
+	capabilities.Validator[TriggerConfig, Input, capabilities.TriggerResponse]
+	connector           connector.GatewayConnector
+	lggr                logger.Logger
+	mu                  sync.Mutex
+	registeredWorkflows map[string]webapiTrigger
+}
+
+var _ capabilities.TriggerCapability = (*triggerConnectorHandler)(nil)
+var _ services.Service = &triggerConnectorHandler{}
+
+func NewTrigger(config string, registry core.CapabilitiesRegistry, connector connector.GatewayConnector, lggr logger.Logger) (*triggerConnectorHandler, error) {
+	if connector == nil {
+		return nil, errors.New("missing connector")
+	}
+	handler := &triggerConnectorHandler{
+		Validator:           capabilities.NewValidator[TriggerConfig, Input, capabilities.TriggerResponse](capabilities.ValidatorArgs{Info: webapiTriggerInfo}),
+		connector:           connector,
+		registeredWorkflows: map[string]webapiTrigger{},
+		lggr:                lggr.Named("WorkflowConnectorHandler"),
+	}
+
+	return handler, nil
+}
+
+// processTrigger iterates over each topic, checking against senders and rateLimits, then starting event processing and responding
+func (h *triggerConnectorHandler) processTrigger(ctx context.Context, gatewayID string, body *api.MessageBody, sender ethCommon.Address, payload webapicapabilities.TriggerRequestPayload) {
+	// Pass on the payload with the expectation that it's in an acceptable format for the executor
+	wrappedPayload, err := values.WrapMap(payload)
+	if err != nil {
+		h.lggr.Errorw("Error wrapping payload")
+		h.sendResponse(ctx, gatewayID, body, webapicapabilities.TriggerResponsePayload{Status: "ERROR", ErrorMessage: "Error wrapping payload"})
+		return
+	}
+	matchedWorkflows := 0
+	var response webapicapabilities.TriggerResponsePayload
+
+	for _, trigger := range h.registeredWorkflows {
+		topics := payload.Topics
+		// empty topics means all topics
+		if len(topics) == 0 {
+			for k := range trigger.allowedTopics {
+				topics = append(topics, k)
+			}
+		}
+
+		for _, topic := range topics {
+			if trigger.allowedTopics[topic] {
+				matchedWorkflows++
+
+				if !trigger.allowedSenders[sender.String()] {
+					h.lggr.Errorw("Unauthorized Sender", "sender", sender.String(), "messageID", body.MessageId)
+					h.sendResponse(ctx, gatewayID, body, webapicapabilities.TriggerResponsePayload{Status: "ERROR", ErrorMessage: "Unauthorized Sender"})
+					return
+				}
+				if !trigger.rateLimiter.Allow(body.Sender) {
+					h.lggr.Errorw("request rate-limited", sender.String(), "messageID", body.MessageId)
+					h.sendResponse(ctx, gatewayID, body, webapicapabilities.TriggerResponsePayload{Status: "ERROR", ErrorMessage: "request rate-limited"})
+					return
```

Comment:
If I send a request with a topic that is meant to kick off multiple workflows, then it will exit out at the first rate limit that happens. This could lead to inconsistent results.

Would it be better to continue on being rate limited? All or nothing?
Maybe the response to the Gateway should be an array of statuses?

Reply:
This is exactly the point I brought up earlier about the issue around boxcarring requests, which is partial failures and status reports.  There's no easy solution imo.  Which is why it's rarely done.

Reply:
I worry about partial success / partial failures.  IMO all or nothing seems to be a simple and acceptable approach here. In the first iteration of the loop,  validate rate limits and allowed senders and collect all triggers that need to be triggered. If any of the validation fails, none of the workflows are triggered. If all validation passes, then start all matching triggers.

one option is to create a ticket for this and come back to it? would make reviewing easier.

Reply:
Same worry.  I decided to skip the ability to invoke multiple topics.  I spoke with Ryan and he's ok with that for V1.  Longer term, the reality is that boxcarring is pretty hard which is why no one really does it.  Because the server can't know what the client wants for boxcarred requests, the client will have to tell it, like "FailFast" or even per condition: {validation: "FailFast", senders: "FailFast", topics: "FailSlow", rateLimit: "FailSlow"}

Reply:
Skip the ability to invoke multiple topics? Or skip the ability to invoke multiple workflows?